### PR TITLE
Migrate build process to a JSON based list

### PR DESCRIPTION
**Pull request by michael-j-green**
This is presented as a proposal to change the build process that builds from a JSON array of elements in a file named ``cores.json`` similar to the below:
```json
{
    "name": "mame2003",
    "extensions": [ "zip" ],
    "makeoptions": {
        "buildpath": "./",
        "makescript": "Makefile",
        "arguments": []
    },
    "options": {
        "file": "MAME 2003 (0.78)/MAME 2003 (0.78).opt",
        "settings": {
            "mame2003_skip_disclaimer": "enabled",
            "mame2003_skip_warnings": "enabled"
        }
    },
    "license": "LICENSE.md",
    "repo": "https://github.com/EmulatorJS/mame2003-libretro"
}
```
> *Notes:*
> * The license attribute in the above is relative to the root of the repo
> * The name should always be the first part of the name of the resultant core .data files

This change doesn't modify the build process to much as the cores are still generated the same way, but it ensures that core configurations are kept with the cores themselves and can be referenced by EJS, with a view to ultimately limiting the statically set core configurations within the EJS code.

There is also a ``build.json`` file who's content looks like this:
```json
{
    "minimumEJSVersion": "4.0.12",
    "version": "2.0.0"
}
```

This build file can be used for version checking to ensure the cores will work with the running version of EJS.

> *Note:* The above build.json file, and the stanza of the ``cores.json`` file relevant to the core is included in the final compressed .data file

The GitHub action has been update to include the above ``cores.json`` file, as well as the ``build.json`` file in the cores.zip file that gets generated, so that they can be deployed to EJS users.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -97,46 +97,32 @@ cd EmulatorJS
...
+    arguments=`echo $(_jq '.') | jq -r '.makeoptions.arguments[] | @base64'`
+
+    argumentstring=""
+    for rowarg in $(echo "${arguments}"); do
+        argumentstring="$argumentstring `echo $rowarg | base64 --decode`"
+    done
+
+    echo "Starting compile of core $name"
+
+    compileProject "$name" "$repo.git" "$buildpath" "$makescript" "$argumentstring"
```
Comment (ethanaobrien) [inline build.sh:119]: Should this be called once per argument string?

File: cores.json [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+        "makeoptions": {
+            "arguments": [],
+            "buildpath": "./",
+            "makescript": "Makefile.libretro"
+        },
+        "options": {},
+        "license": "",
+        "repo": "https://github.com/EmulatorJS/smsplus-gx"
+    }
+]
```
Comment (ethanaobrien) [inline cores.json:466]: (newline?)

File: cores.json [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        "repo": "https://github.com/EmulatorJS/mame2003-libretro"
+    },
+    {
+        "name": "vice_x64",
+        "makeoptions": {
+            "buildpath": "./",
+            "makescript": "Makefile",
+            "arguments": [
+                "EMUTYPE=x64",
+                "SECONDARG=Poo"
```
Comment (ethanaobrien) [inline cores.json:26]: Was this a test?

File: build.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (ethanaobrien) [inline build.json:1]: This would probably require editing the RetroArch build script too, but do you think we could embed this into each core file as `info.json` or something?

File: .github/workflows/compile.yml [hunk 5]
```diff
@@ -13,16 +13,22 @@ jobs:
         uses: actions/checkout@v3
         
       - name: Install Dependencies
-        run: sudo apt install -y p7zip-full binutils-mips-linux-gnu build-essential pkgconf python3 git zip
+        run: sudo apt install -y p7zip-full binutils-mips-linux-gnu build-essential pkgconf python3 git zip jq wget curl
 
       - name: Setup Emscripten
         run: "git clone https://github.com/emscripten-core/emsdk.git .emsdk && cd .emsdk && ./emsdk install latest && ./emsdk activate latest && source ./emsdk_env.sh && wget https://raw.githubusercontent.com/EmulatorJS/build/main/emscripten.patch && patch -u -p0 -i emscripten.patch && cd ../"
```
Comment (ethanaobrien) [inline .github/workflows/compile.yml:19]: We should probably install via my debian repo - since Ive had a broken emscripten commit break this script entirely

File: cores.json [hunk 6]
```diff
@@ -0,0 +1,466 @@
+[
+    {
+        "name": "mame2003-libretro",
+        "makeoptions": {
+            "buildpath": "./",
+            "makescript": "Makefile",
+            "arguments": []
+        },
+        "options": {
```
Comment (ethanaobrien) [inline cores.json:9]: How do we plan on passing these to EmulatorJS? Do you think it would be possible to compress a `default-options.json` (or something) file in the core too? Though this would require editing the dist-cores.sh script in RetroArch

**Review by ethanaobrien — CHANGES_REQUESTED** [review]

File: cores.json (hunk 2 above)
Comment (michael-j-green) [reply cores.json:466]: It shouldn't make a difference, but happy to standardise.

File: .devcontainer/Dockerfile [hunk 7]
```diff
@@ -1,3 +1,3 @@
 FROM ubuntu:latest
 
-RUN apt-get update && apt-get install -y p7zip-full binutils-mips-linux-gnu build-essential pkgconf python3 git zip wget curl libsdl2-dev emscripten
+RUN apt-get update && apt-get install -y p7zip-full binutils-mips-linux-gnu build-essential pkgconf python3 git zip wget curl libsdl2-dev emscripten jq
```
[the remark this answers is not in the document]
Comment (michael-j-green) [reply .devcontainer/Dockerfile:3]: This is building from ubuntu:latest.

We could pin this to a specific Debian version though and import the repo (what's the path to your repo?).

Note that this is just for building cores on a local device when you've cloned the repo - it tells VSCode to create a docker container (if Docker Desktop is installed), which makes it easy to have an isolated development environment - or provide a Linux environment on Windows.

File: cores.json (hunk 2 above)
Comment (ethanaobrien) [reply cores.json:466]: Yeah it wouldn't make a difference. I just try to keep the last line a newline

File: .github/workflows/compile.yml (hunk 5 above)
Comment (michael-j-green) [reply .github/workflows/compile.yml:19]: Agreed... I found an error applying the em patch when I tried building locally yesterday.

File: .devcontainer/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (ethanaobrien) [reply .devcontainer/Dockerfile:3]: Ah I see

https://ethanthesleepy.one/debian/

File: .github/workflows/compile.yml [hunk 8]
```diff
@@ -13,16 +13,22 @@ jobs:
...
 
+      - name: Copy Support Files
+        run: |
+          "cp ./build.json ./compile/EmulatorJS/data/cores/"
+          "cp ./cores.json ./compile/EmulatorJS/data/cores/"
+          "cp ./compile/*-License.txt ./compile/EmulatorJS/data/cores/"
+
       - name: Zip files
-        run: "zip -r 'EmulatorJS Cores.zip' ./compile/EmulatorJS/data/cores/*.data"
+        run: "zip -r 'EmulatorJS Cores.zip' ./compile/EmulatorJS/data/cores/*.data ./compile/EmulatorJS/data/cores/*.json"
```
[the remark this answers is not in the document]
Comment (michael-j-green) [reply .github/workflows/compile.yml:31]: Probably... just wanted to be specific about what's included. Having said that, this is the GitHub action and so would only apply when building via GitHub. But lets be consistent.

File: build.json (hunk 4 above)
Comment (michael-j-green) [reply build.json:1]: This entire file exists to tell EJS what version of the cores package we're using. It would eventually be updated by the GitHub Action - just didn't want to go to that trouble if the changes weren't accepted.

The idea here is that when we create a new cores release on build repo, a GitHub Action would take the version number from the release and update the "version" attribute. Then when EJS is run, it could read this file (which is in the root of the cores.zip) and display the build version in the about box, and also throw an error if the minimumEJSVersion can't be met.

File: build.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
...
+    "minimumEJSVersion": "4.0.12",
+    "version": "2.0.0"
+}
```
[the remark this answers is not in the document]
Comment (michael-j-green) [reply build.json:4]: Done

File: build.sh (hunk 1 above)
Comment (michael-j-green) [reply build.sh:119]: No... the argument strings are the environment variables that get passed to the compileProject command. For example: BUILDARG=1 BUILDARG=2

File: build.sh [hunk 10]
```diff
@@ -97,46 +97,32 @@ cd EmulatorJS
...
+
+    if [ ! -z "$license" -a "$license" != " " ]; then
+        # license url is provided - download it
+        wget "$license?raw=true" -O "./$name-License.txt"
```
[the remark this answers is not in the document]
Comment (michael-j-green) [reply build.sh:123]: I thought about that last night. I still want to leave the license URL in the cores.json file so we can link to it in place. But yes, this wget is sort of redundant. I'll work out a method to copy it rather than download it.

File: cores.json (hunk 6 above)
Comment (michael-j-green) [reply cores.json:9]: This stanza exists to tell EJS to create the .opt files (in the EJS project, see GameManager.js:33). This effectively replaces lines like this: ``this.writeFile("/home/web_user/retroarch/userdata/config/Beetle PSX HW/Beetle PSX HW.opt", 'beetle_psx_hw_renderer = "software"\n');``

When EJS starts, it'll check for the existance of data/cores/cores.json and load it into memory, then search for the selected core - I would then suggest storing the selected core in a variable to make lookups easier.

File: cores.json (hunk 3 above)
Comment (michael-j-green) [reply cores.json:26]: OH MY... :D Yeah, was testing that I could pass multiple arguments.

Cleaned up!

File: build.json (hunk 4 above)
Comment (michael-j-green) [reply build.json:1]: I'll be getting around to attempting to integrate these new json files into EJS in the next few days hopefully. I'll do a similar PR to this one as a proposal.

File: .devcontainer/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (michael-j-green) [reply .devcontainer/Dockerfile]: I've updated the dockerfile to debian:latest, and added your repo.

File: .github/workflows/compile.yml (hunk 5 above)
Comment (michael-j-green) [reply .github/workflows/compile.yml:19]: Resolved this error by pinning the .emsdk version to 3.1.54. That's the last version that the patch works for.

File: build_env.sh [hunk 11]
```diff
@@ -3,6 +3,7 @@
 cd /workspaces/build
 git clone https://github.com/emscripten-core/emsdk.git .emsdk
 cd .emsdk
+git checkout 3.1.54
```
Comment (ethanaobrien) [inline build_env.sh:6]: It's probably safer to run the following commands with `3.1.54` rather than checking out an old tag. I'm not sure how emscripten installs and don't know if it'll lead to package conflicts

**Review by ethanaobrien — CHANGES_REQUESTED** [review]

File: build_env.sh (hunk 11 above)
Comment (michael-j-green) [reply build_env.sh:6]: I'm running a build right now, and it looks like the cores are building ok (no crashes yet!).

In this environment, I've added your repo, used apt-get install emscripten, and then checked out that earlier version of emsdk - seems to be working ok.

Comment (ethanaobrien) [reply build_env.sh:6]: You shouldn’t need to do any checkouts if you install from my repo?

File: .devcontainer/Dockerfile [hunk 12]
```diff
@@ -1,3 +1,7 @@
-FROM ubuntu:latest
+FROM debian:latest
 
-RUN apt-get update && apt-get install -y p7zip-full binutils-mips-linux-gnu build-essential pkgconf python3 git zip wget curl libsdl2-dev emscripten
+RUN apt-get update && apt-get install -y wget curl gpg
+RUN mkdir -p /etc/apt/keyrings
+RUN curl -fsSL https://ethanthesleepy.one/deb-repo/ethanthesleepyone.pgp.key | gpg --dearmor -o /etc/apt/keyrings/ethanthesleepyone.gpg
+RUN echo "deb [signed-by=/etc/apt/keyrings/ethanthesleepyone.gpg] https://ethanthesleepy.one/deb-repo/ nodistro main" | tee /etc/apt/sources.list.d/ethanthesleepyone.list
+RUN apt-get update && apt-get install -y p7zip-full binutils-mips-linux-gnu build-essential pkgconf python3 git zip libsdl2-dev emscripten jq
```
Comment (ethanaobrien) [inline .devcontainer/Dockerfile:7]: Could we add to the readme instructions for people who dont like to use docker for little things like this.

(I can say I am one of those people, I dont even use a "real" IDE. Theyre all bloated and full of spyware nowadays)

File: .gitignore [hunk 13]
```diff
@@ -7,3 +7,5 @@
 !/emscripten.patch
 !/.github
 !/.github/workflows/compile.yml
+!/build.json
+!/cores.json
```
Comment (ethanaobrien) [inline .gitignore:11]: could we add a newline at the end of the file

File: build.json [hunk 14]
```diff
@@ -0,0 +1,4 @@
+{
+    "minimumEJSVersion": "4.0.13",
+    "version": "2.0.0"
+}
```
Comment (ethanaobrien) [inline build.json:4]: Newline please

File: build.json [hunk 15]
```diff
@@ -0,0 +1,4 @@
+{
+    "minimumEJSVersion": "4.0.13",
+    "version": "2.0.0"
```
Comment (ethanaobrien) [inline build.json:3]: Is there a reason we're starting on 2? Not complaining or asking for a change (just yet), just wondering.

File: build.sh [hunk 16]
```diff
@@ -97,65 +97,82 @@ cd EmulatorJS
...
-rm -f *.bc
+    mv core-temp/normal/*.bc ./
+    emmake ./build-emulatorjs.sh emscripten clean no no
+    rm -f *.bc
 
-mv core-temp/legacy/*.bc ./
-emmake ./build-emulatorjs.sh emscripten clean no yes
-rm -f *.bc
+    mv core-temp/threads/*.bc ./
+    emmake ./build-emulatorjs.sh emscripten clean yes no
```
Comment (ethanaobrien) [inline build.sh:144]: Could we dump the output from these emmake calls to a log file instead of to the console?

```bash
rm build.sh
emmake ./build-emulatorjs.sh emscripten clean yes no >> build.sh
```

File: build.sh [hunk 17]
```diff
@@ -97,65 +97,82 @@ cd EmulatorJS
...
+        echo "License file: $name/$license"
+        cp $name/$license "./license.txt"
+    fi
 
-mv core-temp/normal/*.bc ./
-emmake ./build-emulatorjs.sh emscripten clean no no
-rm -f *.bc
+    echo "Build wasm's for core $name"
+    cd "RetroArch"
+    git pull
```
Comment (ethanaobrien) [inline build.sh:136]: are we pulling the "RetroArch" repo with every call to build a core?

File: build.sh [hunk 18]
```diff
@@ -97,65 +97,82 @@ cd EmulatorJS
...
+    if [ ! -z "$license" -a "$license" != " " ]; then
+        # license file is provided - copy it
+        echo "License file: $name/$license"
+        cp $name/$license "./license.txt"
+    fi
 
-mv core-temp/normal/*.bc ./
-emmake ./build-emulatorjs.sh emscripten clean no no
-rm -f *.bc
+    echo "Build wasm's for core $name"
```
Comment (ethanaobrien) [inline build.sh:134]: Should be Building, since it's currently happening (or will happen in a few milliseconds)

File: build.sh [hunk 19]
```diff
@@ -97,65 +97,82 @@ cd EmulatorJS
...
-rm -f *.bc
+    mv core-temp/legacy/*.bc ./
+    emmake ./build-emulatorjs.sh emscripten clean no yes
+    rm -f *.bc
+
+    mv core-temp/legacyThreads/*.bc ./
+    emmake ./build-emulatorjs.sh emscripten clean yes yes
+    rm -f *.bc
+
+    echo "Add support files to $name wasms"
```
Comment (ethanaobrien) [inline build.sh:155]: Could we reword this to be something like `Packing core information for $name`

File: build.sh [hunk 20]
```diff
@@ -97,65 +97,82 @@ cd EmulatorJS
...
+    if [ -f "EmulatorJS/data/cores/$name-thread-legacy-wasm.data" ]; then
+        7z a -t7z EmulatorJS/data/cores/$name-thread-legacy-wasm.data ./core.json ./license.txt ../build.json
+    fi
+
+    # clean up
+    rm ./license.txt
+    rm core.json
+done
 
 cd "$initialPath"
```
Comment (ethanaobrien) [inline build.sh:178]: If no errors occurred, could we delete all the build files? Logs should always be kept.

File: build_env.sh [hunk 21]
```diff
@@ -3,6 +3,7 @@
...
 ./emsdk install latest
 ./emsdk activate latest
```
Comment (ethanaobrien) [inline build_env.sh:8]: Can we specify the verison

```shell
./emsdk install 3.1.54
./emsdk activate 3.1.54
```

File: cores.json [hunk 22]
```diff
@@ -0,0 +1,1016 @@
...
+                "mame2003_skip_disclaimer": "enabled",
+                "mame2003_skip_warnings": "enabled"
+            }
+        },
+        "license": "LICENSE.md",
+        "repo": "https://github.com/EmulatorJS/mame2003-libretro"
+    },
+    {
+        "name": "vice_x64",
+        "extensions": [
```
Comment (ethanaobrien) [inline cores.json:24]: Could we collapse these long blobs of file extensions into one line

File: cores.json [hunk 23]
```diff
@@ -0,0 +1,1016 @@
...
+        "makeoptions": {
+            "buildpath": "./",
+            "makescript": "Makefile",
+            "arguments": []
+        },
+        "options": {},
+        "license": "LICENSE",
+        "repo": "https://github.com/libretro/libretro-crocods"
+    }
+]
```
Comment (ethanaobrien) [inline cores.json:1016]: And finally, newline please

**Review by ethanaobrien — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile (hunk 12 above)
Comment (michael-j-green) [reply .devcontainer/Dockerfile:7]: Absolutely - I hadn't done that part yet. Was waiting for your review :)

Comment (michael-j-green) [reply .devcontainer/Dockerfile:7]: It's already in place :) Look under the ``# Set up your build environment`` and then ``## Local``.

I did update the Readme to say that you may need to run the script as sudo.

File: .gitignore (hunk 13 above)
Comment (michael-j-green) [reply .gitignore:11]: Done

File: build.json (hunk 15 above)
Comment (michael-j-green) [reply build.json:3]: I was working with the idea that the current way things are done would be version 1.0.0. - I'm not married to this assumption though, so happy to reset the number back to 1.0.0 as it's not used anywhere yet.

File: build.json (hunk 14 above)
Comment (michael-j-green) [reply build.json:4]: Done

File: build.sh (hunk 17 above)
Comment (michael-j-green) [reply build.sh:136]: Oh wow... yeah, looks like we are :D

Removed.

File: build.sh (hunk 16 above)
Comment (michael-j-green) [reply build.sh:144]: Compile and emake logs are now sent to a log file named for the core being built.

File: build.sh (hunk 19 above)
Comment (michael-j-green) [reply build.sh:155]: Done

File: build.sh (hunk 20 above)
Comment (michael-j-green) [reply build.sh:178]: Cores now get copied to /output as they're created. Logs are stored in /output/logs. /compile is deleted at the end of the script.

File: build_env.sh (hunk 21 above)
Comment (michael-j-green) [reply build_env.sh:8]: Done

File: cores.json (hunk 23 above)
Comment (michael-j-green) [reply cores.json:1016]: Done

File: cores.json (hunk 22 above)
Comment (michael-j-green) [reply cores.json:24]: VScode decided to help me by cleaning up... 

All cleaned up now.

File: build.sh (hunk 20 above)
Comment (michael-j-green) [reply build.sh:190]: Sample state of the state of the repo after compilation completes. A single core was compiled for this example:
Screenshot 2024-06-28 at 9 15 18 PM

File: .vscode/settings.json
Comment (michael-j-green) [inline .vscode/settings.json:1]: I've included this file to ensure that VScode won't automatically format it.

**Review by ethanaobrien — APPROVED** [review]
